// autoprefixer/lib/selector.js
(function() {
  var Prefixer, Selector, utils,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Prefixer = require('./prefixer');

  utils = require('./utils');

  Selector = (function(_super) {
    __extends(Selector, _super);

    function Selector(name, prefixes, all) {
      this.name = name;
      this.prefixes = prefixes;
      this.all = all;
      this.regexpCache = {};
    }

    Selector.prototype.check = function(rule, prefix) {
      var name;
      name = prefix ? this.prefixed(prefix) : this.name;
      if (rule.selector.indexOf(name) !== -1) {
        return !!rule.selector.match(this.regexp(prefix));
      } else {
        return false;
      }
    };

    Selector.prototype.checker = function(prefix) {
      var _this = this;
      return function(rule) {
        return _this.check(rule, prefix);
      };
    };

    Selector.prototype.prefixed = function(prefix) {
      return this.name.replace(/^([^\w]*)/, '$1' + prefix);
    };

    Selector.prototype.regexp = function(prefix) {
      var name;
      if (this.regexpCache[prefix]) {
        return this.regexpCache[prefix];
      }
      name = prefix ? this.prefixed(prefix) : this.name;
      return this.regexpCache = new RegExp('(^|[^:])' + utils.escapeRegexp(name), 'gi');
    };

    Selector.prototype.replace = function(selector, prefix) {
      return selector.replace(this.regexp(), '$1' + this.prefixed(prefix));
    };

    Selector.prototype.add = function(rule, prefix) {
      var cloned, prefixed;
      prefixed = this.replace(rule.selector, prefix);
      if (rule.parent.some(function(i) {
        return i.selector === prefixed;
      })) {
        return;
      }
      cloned = this.clone(rule, {
        selector: prefixed
      });
      return rule.parent.insertBefore(rule, cloned);
    };

    return Selector;

  })(Prefixer);

  module.exports = Selector;

}).call(this);
